specs/src/shrex/client-server.md

```diff
@@ -0,0 +1,388 @@
+# SHREX Client/Server Specification
+
+## Abstract
+
+This specification defines a request-response protocol that enables nodes to retrieve data from peers. The protocol provides a simple interface for requesting specific data and receiving responses.
+
+## Overview
+
+It implements a client-server architecture for data exchange:
+
+- **SHREX Client**: Makes requests to retrieve data from a specific peer
+- **SHREX Server**: Responds to requests by serving data from local storage
+
+## Transport Protocol
+
+It uses libp2p streaming as its transport layer.
+
+### Protocol ID
+
+The protocol ID follows the format:
+
+```text
+/<network-id>/shrex/v0.1.0/<request-name>
+```
+
+Where:
+
+- `<network-id>`: Network identifier
+- `/shrex/v0.1.0/`: Protocol name and version
+- `<request-name>`: Type of request being made (e.g., "row", "sample", "eds")
+
+### Stream Lifecycle
+
+1. **Stream Opening**: Client opens a new stream to the server using the protocol ID
+2. **Request Phase**: Client writes request and closes the write side
+3. **Response Phase**: Server sends status message
+   - If status is `OK`: Server sends data payload, then closes stream
+   - If status is not `OK`: Server closes stream immediately
+4. **Stream Closing**: Stream is closed after response is complete
+
+### Timeouts
+
+Both client and server implement timeouts to protect resources:
+
+- **Read Timeout**: Maximum time to read data from stream
+- **Write Timeout**: Maximum time to write data to stream
+- **Handle Request Timeout**: Maximum time for server to process request
+
+## Message Format
+
+### Encoding
+
+Request and response messages use different encoding schemes:
+
+- **Request Messages**: Binary encoded using `MarshalBinary()` method
+- **Response Messages**: Protocol Buffers (protobuf) encoded
+
+### Request Messages
+
+Request messages are binary encoded.
+
+The request is serialized to binary format and written directly to the stream. Each request type includes:
+
+- Height: Block height for the requested data
+- Type-specific parameters
+
+### Response Messages
+
+#### Status Message
+
+Every server response begins with a status message:
+
+```protobuf
+enum Status {
+   INVALID = 0;      // Invalid/unknown status
+   OK = 1;           // Data found and will be sent
+   NOT_FOUND = 2;    // Requested data not found
+   INTERNAL = 3;     // Internal server error
+}
+
+message Response {
+   Status status = 1;
+}
+```
```

Comment:
After handling requests server must send Status message indicating result of handling.